WordPress/Classes/Services/MediaService.m

```diff
@@ -56,6 +60,141 @@ - (id)initWithManagedObjectContext:(NSManagedObjectContext *)context
     return self;
 }
 
+- (void)syncMediaLibraryForBlog:(Blog *)blog
+                        success:(void (^)())success
+                        failure:(void (^)(NSError *error))failure
+{
+    id<MediaServiceRemote> remote = [self remoteForBlog:blog];
+    [remote getMediaLibraryForBlog:blog
+                           success:^(NSArray *media) {
+                               [self.managedObjectContext performBlock:^{
+                                   [self mergeMedia:media forBlog:blog completionHandler:success];
+                               }];
+                           }
+                           failure:^(NSError *error) {
+                               if (failure) {
+                                   [self.managedObjectContext performBlock:^{
+                                       failure(error);
+                                   }];
+                               }
+                           }];
+}
+
+- (void)mergeMedia:(NSArray *)media forBlog:(Blog *)blog completionHandler:(void (^)(void))completion
+{
+    NSMutableArray *mediaToKeep = [NSMutableArray array];
+    for (RemoteMedia *remote in media) {
+        Media *local = [self findMediaWithID:remote.mediaID inBlog:blog];
+        if (!local) {
+            local = [self newMediaForBlog:blog];
+            local.remoteStatus = MediaRemoteStatusSync;
+        }
+        [self updateMedia:local withRemoteMedia:remote];
+        [mediaToKeep addObject:local];
+    }
+    
+    NSSet *existingMedia = blog.media;
+    NSMutableOrderedSet *mediaToThumbnail = [NSMutableOrderedSet orderedSet];
+    if (existingMedia.count > 0) {
+        for (Media *existing in existingMedia) {
+            if (![mediaToKeep containsObject:existing] && existing.remoteURL != nil) {
+                [self.managedObjectContext deleteObject:existing];
+            } else if (existing.remoteURL.length && !existing.thumbnail && existing.mediaType == MediaTypeImage) {
+                [mediaToThumbnail addObject:existing];
+            }
+        }
+    }
+    
+    [[ContextManager sharedInstance] saveDerivedContext:self.managedObjectContext];
+    
+    [mediaToThumbnail sortUsingComparator:^NSComparisonResult(Media *left, Media *right) {
+        return [right.creationDate compare:left.creationDate];
+    }];
+    [self getThumbnailsForMedia:mediaToThumbnail];
```

Comment:
I'd generally agree that loading only if ever displayed is preferred behaviour, yes. The task conditions include "Make media available for composing while network is not available" -- and I interpreted that to mean that all thumbnails should be loaded as soon as their Media existence was known. Perhaps we could limit automatic thumbnail loading to the latest 100 additions? I don't know where I would find data to prove or disprove this, but if other people blog the way I do they don't refer to images older than that very often, so that would probably address the task's victory conditions and data gobbler concerns for the general case.